### PR TITLE
[4.0] update the custom elements

### PR DESCRIPTION
**Pull request by dgrammatiko**
Pull Request for Issue # .

### Summary of Changes
- update webcomponentjs (the polyfills) to the latest version (2.0.3)
- remove the hacked tabs custom element from media/system/webcomponents
- Fix the two known bugs for tabs (appendChild instead of append and tabs.id instead of tabs.hash)

### Testing Instructions
Apply patch (probably won't work with patchtester)
Edit an article
Check the tabs functionality
Save Check the alert element
exit and re-edit the same article, observe that the tabstate is kept

### Documentation Changes Required

@wilsonge

### CODE REVIEW

**Comment by astridx** [comment]
I have tested this item :white_check_mark: successfully on 56722dcd2423455207223f8f6cc14eb2e583ba7a<br /><br />Before and after I applied the patch I could edit an article and if I check the tabs functionality everything was OK. 
I closed and re-edit the same article - the tabstate is kept.


I spotted that the file /media/vendor/joomla-custom-elements/js/joomla-tab.js should not be minimized. And it would be great if you  could explain why this files for joomla-tab are in the folder /media/vendor/joomla-custom-elements/js and not in the folder /media/system/webcomponents/js<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/21204">issues.joomla.org/tracker/joomla-cms/21204</a>.</sub>

**Comment by brianteeman** [comment]
Tested in chrome - all good
Tested in IE11 - the tabs are present now BUT now the editor doesnt load
> : Unable to get property 'nodeName' of undefined or null reference in tinymce.min.js

**Comment by astridx** [comment]
I have tested on firefox 61.0.1 (64-Bit) on Ubuntu 18.04

**Comment by brianteeman** [comment]
I have tested this item :red_circle: unsuccessfully on 56722dcd2423455207223f8f6cc14eb2e583ba7a<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/21204">issues.joomla.org/tracker/joomla-cms/21204</a>.</sub>

**Comment by anuragteapot** [comment]
I have tested this item :red_circle: unsuccessfully on 56722dcd2423455207223f8f6cc14eb2e583ba7a<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/21204">issues.joomla.org/tracker/joomla-cms/21204</a>.</sub>

**Comment by anuragteapot** [comment]
@dgrammatiko Few things I noticed

1. The issue still present.
screenshot from 2018-07-21 09-12-36

2. In debug mode why its loading `.min.js` files ?
selection_013

3. Is these custom elements are not a part of joomla ? 
Because, these are kept  in `media/vendor` ?

File: media/vendor/joomla-custom-elements/js/joomla-tab-es5.js [hunk 1]
```diff
@@ -0,0 +1,486 @@
...
+        if (this.hasNested) {
+          // @todo use the recall attribute
+          var _href = sessionStorage.getItem(this.getStorageKey());
+          if (_href) {
+            tabLinkHash.push(_href);
+          }
+          // @todo end
+
+          // Add possible parent tab to the aray for activation
+          console.log(tabLinkHash);
```
Comment (anuragteapot) [inline media/vendor/joomla-custom-elements/js/joomla-tab-es5.js]: remove console.log()

File: media/vendor/joomla-custom-elements/js/joomla-tab.js [hunk 2]
```diff
@@ -0,0 +1 @@
...ak;case 39:case 40:r.querySelector("a").click(),r.querySelector("a").focus(),t.preventDefault()}})}getStorageKey(){return window.location.href.toString().split(window.location.host)[1].replace(/&return=[a-zA-Z0-9%]+/,"").split("#")[0]}saveState(t){const e=this.getStorageKey();sessionStorage.setItem(e,t)}checkView(t){const e=t.querySelector("ul");let i=[];if(document.body.getBoundingClientRect().width>920){if("tabs"===this.view)return;t.view="tabs";const r=[].slice.call(e.querySelectorAll("section"));for(let e=0,s=r.length;e<s;++e)r[e].parentNode.parentNode.parentNode===t&&i.push(r[e]);i.length&&i.forEach(e=>{t.appendChild(e)})}else{if("accordion"===this.view)return;t.view="accordion";const e=[].slice.call(t.querySelectorAll("section"));for(let r=0,s=e.length;r<s;++r)e[r].parentNode===t&&i.push(e[r]);i.length&&i.forEach(e=>{const i=t.querySelector('a[aria-controls="'+e.id+'"]');i.parentNode.parentNode===t.firstElementChild&&i.parentNode.appendChild(e)})}}findAncestor(t,e){for(;(t=t.parentElement)&&t.nodeName.toLowerCase()!==e;);return t}dispatchCustomEvent(t,e,i){const r=new CustomEvent(t,{bubbles:!0,cancelable:!0});r.relatedTarget=i,e.dispatchEvent(r),e.removeEventListener(t,e)}});
```
Comment (anuragteapot) [inline media/vendor/joomla-custom-elements/js/joomla-tab.js]: minified version in `.js` ?

**Comment by anuragteapot** [comment]
@dgrammatiko As I can see.
If you test this on Article in your test case. It's working fine. Because Its going through first case 

screenshot from 2018-07-21 09-35-14

`this.hasNested` is true for the article. 
that's why its giving output in the console. 

screenshot from 2018-07-21 09-38-27

And, in mine case, I tested at `com_templates` and it's going through 2nd case `else` .

File: media/vendor/joomla-custom-elements/js/joomla-tab-es5.js [hunk 3]
```diff
@@ -0,0 +1,486 @@
...
+            if (tab.parentNode === self) {
+              tabsEl.push(tab);
+            }
+          });
+        } else {
+          // Activate the correct tab
+          tabs.forEach(function (tab) {
+            if (tabLinkHash.length) {
+              var theId = '#tab-' + tab.id;
+              if (tabLinkHash.indexOf(theId) > -1) {
```
Comment (anuragteapot) [inline media/vendor/joomla-custom-elements/js/joomla-tab-es5.js]: May be if we set  `if (tabLinkHash.indexOf(theId) ===  -1)`  it works ?

**Comment by dgrammatiko** [comment]
> Tested in IE11 - the tabs are present now BUT now the editor doesnt load

Well, Houston we have a problem. I'm trying to figure out why this is happening

@Anu1601CS probably the conditional is wrong, will take a look at that once I fix the polyfills situation, which is way more severe

**Comment by anuragteapot** [comment]
@dgrammatiko Ok, Thanks

**Comment by dgrammatiko** [comment]
FWIW I've already filled an issue: https://github.com/tinymce/tinymce/issues/4502 as all the combinations of the polyfills that include the shadowDOM will fail. This is a tinyMCE issue

**Comment by dgrammatiko** [comment]
About the tinyMCE problem: it's gonna be fixed upstream.
https://github.com/tinymce/tinymce/issues/4502#issuecomment-407412557

**Comment by fuzzbomb** [comment]
@brianteeman asked me to look here.

I'm pretty new to Joomla development - I've got it running locally, but the JS files loaded by the browser are all minified versions.  Is there a way to get Joomla to use the original versions?

**Comment by laoneo** [comment]
Change to debug mode in the Joomla configuration.

**Comment by brianteeman** [comment]
@fuzzbomb send me a message if you get stuck.

**Comment by infograf768** [comment]
I have tested this item :white_check_mark: successfully on 8820ffa32bcc3ec7a8247d63f62b606ebff61239<br /><br />All tested fine.

We recover the tabs (yes, including in Firefox) although their display have changed (certainly a matter of css)

screen shot 2018-08-03 at 08 15 20


Did not test with IE as I use mac.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/21204">issues.joomla.org/tracker/joomla-cms/21204</a>.</sub>

**Comment by ghost** [comment]
@Anu1601CS can you please retest?

**Comment by infograf768** [comment]
Needs npm to be tested

**Comment by ghost** [comment]
npm = ?

**Comment by TobsBobs** [comment]
npm = Node Package Manager

**Comment by ghost** [comment]
Thanks @TobsBobs, i know NPM as "New Public Management" :-)

**Comment by wojsmol** [comment]
@franz-wohlkoenig In simple words - needs a little knowledge about you Mac Terminal. Can be referred also as `node` or `nodee.js`

**Comment by wilsonge** [comment]
Given we've documented the tinymce issue at https://github.com/joomla/joomla-cms/issues/21374 I'm going to merge this as it's a 'safe' update and at least moves ff one step forward

**Pull request merged**